tonatona-sample/src/Tonatona/Sample.hs

```diff
@@ -133,22 +134,56 @@ instance TonaServant.HasConfig Config where
 -- Shared
 
 data Shared = Shared
-  { tonaDb :: TonaDb.Shared
+  { tonaDb :: SharedDb
   , tonaLogger :: TonaLogger.Shared
   }
 
+
+data SharedDb
+  = SharedSqlite TonaDbSqlite.Shared
+  | SharedPostgres TonaDbPostgres.Shared
+
+sharedDbRun ::
+     (TonaDbPostgres.HasShared shared, TonaDbSqlite.HasShared shared)
+  => ReaderT SqlBackend (TonaM Config shared) a
+  -> TonaM Config shared a
+sharedDbRun query = do
+  conf <- askConf
+  case dbToUse conf of
+    Sqlite -> TonaDbSqlite.run query
+    PostgreSQL -> TonaDbPostgres.run query
+
+sharedDbMigrate ::
+     (TonaDbPostgres.HasShared shared, TonaDbSqlite.HasShared shared)
+  => Migration
+  -> TonaM Config shared ()
+sharedDbMigrate migration = do
+  conf <- askConf
+  case dbToUse conf of
+    Sqlite -> TonaDbSqlite.runMigrate migration
+    PostgreSQL -> TonaDbPostgres.runMigrate migration
+
 instance Plug Config Shared where
   init conf = do
     let db =
           case dbToUse conf of
-            Sqlite -> TonaDbSqlite.init conf stdoutLogger
-            PostgreSQL -> TonaDbPostgres.init conf stdoutLogger
+            Sqlite -> SharedSqlite <$> TonaDbSqlite.init conf stdoutLogger
+            PostgreSQL -> SharedPostgres <$> TonaDbPostgres.init conf stdoutLogger
     Shared
       <$> db
       <*> TonaLogger.init stdoutLogger
 
-instance TonaDb.HasShared Shared where
-  shared = tonaDb
+instance TonaDbSqlite.HasShared Shared where
+  shared s =
+    case (tonaDb :: Shared -> SharedDb) s of
+      SharedSqlite sqliteShared -> sqliteShared
+      _ -> error "This is not an sqlite shared type."
+
+instance TonaDbPostgres.HasShared Shared where
+  shared s =
+    case (tonaDb :: Shared -> SharedDb) s of
+      SharedPostgres postgresShared -> postgresShared
+      _ -> error "This is not a postgres shared type."
```

Comment:
Since `HasShared` is no longer defined in `Tonatona.Db`, but instead now defined separately in both `Tonatona.Db.Postgresql` and `Tonatona.Db.Sqlite`, the user is required to create a separate instance of each of these `HasShared` classes.

Unfortunately, these classes call `error` at runtime.  This seems like an unfortunate design decision.  I can't think of any way to get around this.